_posts/2019-02-11-configurable-builds-part-1.md

```diff
@@ -0,0 +1,284 @@
+doesn't even matter if no one's ever heard of your platform.
+
+With the new API, our example can be rewritten as:
+
+```python
+$ cat platforms/BUILD
+constraint_setting(name = "os")
+constraint_value(name = "android", constraint_setting = ":os")
+constraint_value(name = "linux", constraint_setting = ":os")
+
+constraint_setting(name = "device")
+constraint_value(name = "phone", constraint_setting = ":device")
+constraint_value(name = "tablet", constraint_setting = ":device")
+
+platform(name = "pixel3", constraint_values = [":android", ":phone"])
+```
+
+```python
+$ cat helpers/BUILD
+config_setting(name = "android", constraint_values = ["//platforms:android"])
+java_library(
+    name = "supporting_library",
+    deps = select({
+        ":android": [":extra_required_android_deps"],
+        "//conditions:default": []
+    }))
+```
+
+and built with:
+
+    $ bazel build //my:binary --platforms=//platforms:pixel3 --define MYFEATURE=1
+
+This is a significant improvement over before. It's more concise and you no
+longer have to care what the CPU is. *Any* platform with the `constraint_value`
+`":android"` is an Android platform. So it's easy to express *exactly* what you
+want and have `select`, rules, and toolchains understand it the same way.
+`MYFEATURE=1`, which isn't a platform property, remains as before.
+
+Rules understand `":android"` by declaring [toolchain types](https://docs.bazel.build/versions/master/toolchains.html#writing-rules-that-use-toolchains).
+For example, `java_binary` can be defined in Starlark as
+
+```python
+java_binary = rule(..., toolchains = ["@bazel_tools//tools/jdk:toolchain_type"])
+```
+
+This tells Bazel that `java_binary` understands toolchains that set
+`["@bazel_tools//tools/jdk:toolchain_type"]`. You can then write a Java
+toolchain for Android as:
+
+```python
+toolchain(
+    name = "android_jdk",
+    toolchain_type = ["@bazel_tools//tools/jdk:toolchain_type"],
+    target_compatible_with = ["//platforms:android"],
+    toolchain = ":android_jdk_provider_rule")
+```
+
+where `android_jdk_provider_rule` is a Starlark rule that
+[provides](https://docs.bazel.build/versions/master/skylark/rules.html#providers)
+ access to the actual JDK tools.
+
+This takes some infrastructure to set up, but the result is magic. Call your
+build with `--platforms=//platforms:pixel3` and `java_binary` automatically
+uses `":android_jdk"`. All rules can join in on this. The only obligations rule
+designers have are to write [Starlark rules](https://docs.bazel.build/versions/master/toolchains.html#defining-toolchains)
+describing how their toolchains work and define toolchains for the platforms
+they support.
+
+This brings us closer to the goal of `$ bazel build //:all` *just working* by
+replacing ad hoc language-specific flags with a single flag that works everywhere.
+
+### Status
+
+You can use platforms today. The catch is that rules have to opt in support by
```

Comment:
I'd like to see "You can use platforms today" be a little clearer—I don't think you mean literal platforms, but rather that you can configure Bazel to have per-target platforms today (or something similar). Is that right?

Reply:
Can you clarify?

"You can configure Bazel to have per-target platforms" is something I'm very much not saying here. Part 2 gets more into those themes. But we're not really pushing that narrative yet.

Reply:
Ah yeah, sure thing, sorry for the confusing feedback.

I *think* when you say the word "platforms" in this sentence, you're talking about something that isn't a literal platform—I think you're using it as shorthand for "You can configure builds to target specific platforms using Bazel today" or something like that. If that is the case, it would be good to clarify that shorthand (because most readers won't be so close to the problem that they understand how you're using the term). If you do mean literal platforms (e.g., I could substitute the word "macOS" in that sentence and get a similar meaning), then I misunderstood, and you can resolve this comment.

Does that help?

Reply:
Ok, tried to be more clear.